nft/ethereum_payment_obligation.go

```diff
@@ -267,40 +287,59 @@ type MintRequest struct {
 }
 
 // NewMintRequest converts the parameters and returns a struct with needed parameter for minting
-func NewMintRequest(tokenID TokenID, to common.Address, anchorID anchors.AnchorID, proofs []*proofspb.Proof, rootHash [32]byte) (MintRequest, error) {
-	proofData, err := createProofData(proofs)
+func NewMintRequest(model documents.Model, tokenID TokenID, to common.Address, anchorID anchors.AnchorID, nextAnchorID anchors.AnchorID, proofs []*proofspb.Proof, rootHash [32]byte) (MintRequest, error) {
+	proofData, err := createProofData(model, proofs)
 	if err != nil {
 		return MintRequest{}, err
 	}
 
 	return MintRequest{
-		To:         to,
-		TokenID:    tokenID.BigInt(),
-		TokenURI:   tokenID.URI(),
-		AnchorID:   anchorID.BigInt(),
-		MerkleRoot: rootHash,
-		Values:     proofData.Values,
-		Salts:      proofData.Salts,
-		Proofs:     proofData.Proofs}, nil
+		To:           to,
+		TokenID:      tokenID.BigInt(),
+		TokenURI:     tokenID.URI(),
+		AnchorID:     anchorID.BigInt(),
+		NextAnchorID: nextAnchorID.BigInt(),
+		Props:        proofData.Props,
+		Values:       proofData.Values,
+		Salts:        proofData.Salts,
+		Proofs:       proofData.Proofs}, nil
 }
 
 type proofData struct {
+	Props  [][]byte
 	Values [][]byte
 	Salts  [][32]byte
 	Proofs [][][32]byte
 }
 
-func createProofData(proofspb []*proofspb.Proof) (*proofData, error) {
-	var values = make([][]byte, len(proofspb))
+func createProofData(model documents.Model, proofspb []*proofspb.Proof) (*proofData, error) {
+	// TODO cleanup hard coded indexes using the model props
+	readRoleIndex := 5
+	tokenRoleIndex := 7
+	var props = make([][]byte, 2)  // props are only required for readRole.property, tokenRole.property
+	var values = make([][]byte, 4) // values are only required for readRole.Value
 	var salts = make([][32]byte, len(proofspb))
 	var proofs = make([][][32]byte, len(proofspb))
 
 	// TODO remove later
 	//proof, _ := documents.ConvertDocProofToClientFormat(&documents.DocumentProof{FieldProofs: proofspb})
 	//log.Info(json.MarshalIndent(proof, "", "  "))
-
 	for i, p := range proofspb {
-		values[i] = p.Value
+		if i == readRoleIndex {
+			props[0] = p.GetCompactName()
+		}
+		if i == tokenRoleIndex {
+			props[1] = p.GetCompactName()
+		}
+
+		if i < 3 {
+			values[i] = p.Value
+		}
+
+		if i == readRoleIndex {
```

Comment:
you are already doing the same check above, move the logic there instead

Reply:
thanks, missed that. But this is something Im going to refactor soon as commented